optuna/samplers/_cmaes.py

```diff
@@ -325,23 +353,31 @@ def sample_relative(
     def _restore_optimizer(
         self,
         completed_trials: "List[optuna.trial.FrozenTrial]",
-    ) -> Tuple[Optional[CMA], int]:
+    ) -> Tuple[Optional[CmaClass], int]:
+        if not self._use_separable_cma:
+            attr_key_optimizer = "cma:optimizer"
+            attr_key_n_restarts = "cma:n_restarts"
+        else:
+            attr_key_optimizer = "sepcma:optimizer"
+            attr_key_n_restarts = "sepcma:n_restarts"
+
         # Restore a previous CMA object.
         for trial in reversed(completed_trials):
             optimizer_attrs = {
                 key: value
                 for key, value in trial.system_attrs.items()
-                if key.startswith("cma:optimizer")
+                if key.startswith(attr_key_optimizer)
             }
             if len(optimizer_attrs) == 0:
                 continue
 
-            # Check "cma:optimizer" key for backward compatibility.
-            optimizer_str = optimizer_attrs.get("cma:optimizer", None)
-            if optimizer_str is None:
+            if not self._use_separable_cma and "cma:optimizer" in optimizer_attrs:
+                # Check "cma:optimizer" key for backward compatibility.
+                optimizer_str = optimizer_attrs["cma:optimizer"]
+            else:
                 optimizer_str = _concat_optimizer_attrs(optimizer_attrs)
```

Comment:
[Question] Do you expect users to switch CMA-ES to sep CMA-ES in the middle of the optimization?
For example, the following code does not raise any errors or warnings, but I'm not sure if it works properly or not.

```python
import optuna

def objective(trial):
    x = trial.suggest_float("x", 0, 1)
    y = trial.suggest_float("y", 0, 1)

    return (x - 0.5) ** 2 + (y - 0.5) ** 2

study = optuna.create_study(sampler=optuna.samplers.CmaEsSampler())
study.optimize(objective, n_trials=20)

study.sampler=optuna.samplers.CmaEsSampler(use_separable_cma=True)
study.optimize(objective, n_trials=20)

```

Reply:
Yes, I expected such a situation and this PR supports the case you commented.

Although CmaEsSampler stores its `optimizer` and the restart count in `trial_system_attr`, the different system_attr keys are used between (normal) CMA-ES and sep-CMA-ES. So the parameters of sep-CMA-ES is initialized with `mean=np.array([0.5, 0.5])` and `sigma=(1-0)/6`. It means sep-CMA-ES does not exploit the state of (normal) CMA-ES.

Reply:
Thank you for your explanation. That makes sense.
Users may ask questions about this behavior, and I'll answer them with the reference to your comment.